### ISSUE TITLE
MqttClient disconnected event not trigger in 2.8.0 and 2.8.1

### ISSUE THREAD

**Issue by iamoldli** [chkr1011/MQTTnet#358]
I find the ```WaitForTaskAsync``` method in ```MqttClient.cs```  is always deadlock

```
 private static async Task WaitForTaskAsync(Task task, Task sender)
        {
            if (task == sender || task == null)
            {
                return;
            }

            if (task.IsCanceled || task.IsCompleted || task.IsFaulted)
            {
                return;
            }

            try
            {
                **await task.ConfigureAwait(false);**
            }
            catch (OperationCanceledException)
            {
            }
        }
```

**Comment by JanEggers** [comment]
can you verify the changes in #343?

**Comment by iamoldli** [comment]
I've found the problem,when the server is closed, the client's method ```SendKeepAliveMessagesAsync``` and ```ReceivePacketsAsync``` will be at the same time catch exception and will call the ```DisconnectInternalAsync``` method,but in under code will be deadlock,
```
await WaitForTaskAsync(_packetReceiverTask, sender).ConfigureAwait(false);
await WaitForTaskAsync(_keepAliveMessageSenderTask, sender).ConfigureAwait(false);
```
Now,I add under code,the client's event ```Disconnected``` will be call

```
private bool _isDisconnecting;
        private async Task DisconnectInternalAsync(Task sender, Exception exception)
        {
            lock (_disconnectLock)
            {
                if (_isDisconnecting)
                    return;

                _isDisconnecting = true;
            }

            try
            {
                if (IsConnected)
                {
                    IsConnected = false;

                    InitiateDisconnect();

                    await WaitForTaskAsync(_packetReceiverTask, sender).ConfigureAwait(false);
                    await WaitForTaskAsync(_keepAliveMessageSenderTask, sender).ConfigureAwait(false);

                    if (_adapter != null)
                    {
                        await _adapter.DisconnectAsync(_options.CommunicationTimeout, CancellationToken.None)
                            .ConfigureAwait(false);
                    }

                    _logger.Verbose("Disconnected from adapter.");
                }
            }
            catch (Exception adapterException)
            {
                _logger.Warning(adapterException, "Error while disconnecting from adapter.");
            }
            finally
            {
                _adapter?.Dispose();
                _adapter = null;
                _cancellationTokenSource?.Dispose();
                _cancellationTokenSource = null;
                _cleanDisconnectInitiated = false;

                _isDisconnecting = false;

                _logger.Info("Disconnected.");

                Disconnected?.Invoke(this, new MqttClientDisconnectedEventArgs(IsConnected, exception));
            }
        }

```

My English is not good,Can you verify it for me?

**Comment by PitySOFT** [comment]
I can confirm this issue.

**Comment by JTrotta** [comment]
Me too. This issue is confirmed on Mono.

**Comment by eeybye** [comment]
Can also confirm this issue for version 2.8.1 on net core/standard 2.0.
When the client then reconnects, the disconnect event is fired after the connect event.

**Comment by josefbicik** [comment]
Hi, the problem seems to be fixed in 2.8.2-rc1. The client correctly detects disconnection and reconnects without any problems to Mosquitto broker.
